nncf/experimental/onnx/graph/nncf_graph_builder.py

```diff
@@ -0,0 +1,136 @@
+"""
+ Copyright (c) 2022 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from onnx import ModelProto  # pylint: disable=no-name-in-module
+
+from nncf.common.graph import NNCFGraph
+from nncf.common.graph.definitions import NNCFGraphNodeType
+from nncf.common.graph.layer_attributes import Dtype
+
+from nncf.experimental.onnx.graph.onnx_graph import ONNXGraph
+from nncf.experimental.onnx.graph.metatypes.onnx_ops import ONNX_OPERATION_METATYPES
+from nncf.experimental.onnx.graph.metatypes.onnx_ops import ConstantMetatype
+
+
+class GraphConverter:
+    """
+    Builds the NNCFGraph from an ONNX model
+    """
+
+    @staticmethod
+    def create_nncf_graph(onnx_model: ModelProto) -> NNCFGraph:
+        """
+        Adds all ONNX nodes from 'onnx_model' and then adds thr special input_nodes and output_nodes.
+        """
+
+        nncf_graph = NNCFGraph()
+        onnx_graph = ONNXGraph(onnx_model)
+        for node in onnx_graph.get_all_nodes():
+            node_name = node.name
+            node_type = node.op_type
+            metatype = ONNX_OPERATION_METATYPES.get_operator_metatype_by_op_name(node_type)
+            if metatype == ConstantMetatype:  # We don't need to quantize Constants
+                continue
+            nncf_graph.add_nncf_node(node_name=node_name,
+                                     node_type=node_type,
+                                     node_metatype=metatype,
+                                     layer_attributes=None)
+        input_counter = {}
+        output_counter = {}
+        for output_node in nncf_graph.get_all_nodes():
+            output_node_id = output_node.node_id
+            outputs = onnx_graph.get_all_node_outputs(output_node.node_name)
+            for output in outputs:
+                nodes = onnx_graph.get_nodes_by_input(output)
+                shape = onnx_graph.get_node_output_shape(output)
+                for in_node in nodes:
+                    in_node_id = nncf_graph.get_node_by_name(in_node.name).node_id
+                    input_counter[in_node_id] = input_counter.get(in_node_id, -1) + 1
+                    output_counter[output_node_id] = input_counter.get(output_node_id, -1) + 1
```

Comment:
Use `defaultdict(int)` for your counter dicts and you won't have to keep this awkward code.

Reply:
Done